File: frontend/src/features/Train/types/trainTypes.ts
import { DATA_SOURCE_ARR } from "../constants/trainConstants";

// keep in sync with schemas.py
export type DATA_SOURCE = typeof DATA_SOURCE_ARR[number];

export type TRAIN_STATUS =
  | "QUEUED"
  | "STARTING"
  | "UPLOADING"
  | "TRAINING"
  | "SUCCESS"
  | "ERROR";

export interface BaseTrainspaceData {
  name: string;
  dataSource: DATA_SOURCE;
  step: number;
}

// basic information, used on dashboard
export interface TrainResultsData {
  name: string;
  trainspaceId: string;
  dataSource: DATA_SOURCE;
  status: TRAIN_STATUS;
  created: Date;
  uid: string;
}

export type CHART_TYPE = "LINE" | "AUC/ROC" | "CONFUSION_MATRIX"

export type Chart = TimeSeriesChart | AucRocChart | ConfusionMatrixChart

export interface TimeSeriesMetric {
  x_name: string;
  y_name: string;

  x_values: number[];
  y_values: number[];
}

export interface TimeSeriesChart {
  name: string;

  time_series: TimeSeriesMetric[]
  chart_type: "LINE" 
  graph_index: number;
}

export interface AucRocChart {
  name: string;

  values: [number[], number[], number][];

  chart_type: "AUC/ROC"
  graph_index: number;
}

export interface ConfusionMatrixChart {
  name: string;
  
  values: number[][];

  chart_type: "CONFUSION_MATRIX"
  graph_index: number;
}

// more detailed information, used when viewing a run
export interface DetailedTrainResultsData {
  basic_info: TrainResultsData

  all_metrics: Chart[]
}

export interface FileUploadData {
  name: string;
  lastModified: string;
  contentType: string;
  sizeInBytes: number;
}

export interface DatasetData {
  isDefaultDataset: boolean;
  name: string;
}

File: training/training/core/celery/train_types.py
import datetime
from dataclasses import dataclass
from typing import List, Literal, Tuple, Union

from ninja import Schema

# keep in sync with trainTypes.ts
DATA_SOURCE = Literal[
    "TABULAR",
    "PRETRAINED",
    "IMAGE",
    "AUDIO",
    "TEXTUAL",
    "CLASSICAL_ML",
    "OBJECT_DETECTION",
]

TRAIN_STATUS = Literal[
    "QUEUED", "STARTING", "UPLOADING", "TRAINING", "SUCCESS", "ERROR"
]


class TrainResultsData(Schema):
    name: str
    trainspaceId: str
    dataSource: DATA_SOURCE
    status: TRAIN_STATUS
    created: datetime.date
    step: str
    uid: str


CHART_TYPE = Literal["LINE", "AUC/ROC", "CONFUSION_MATRIX"]


class TimeSeriesMetric(Schema):
    x_name: str
    y_name: str

    x_values: List[float]
    y_values: List[float]


class TimeSeriesChart(Schema):
    name: str

    time_series: List[TimeSeriesMetric]
    chart_type = "LINE"
    graph_index: int


class AucRocChart(Schema):
    name: str

    values: List[Tuple[List[float], List[float], float]]
    chart_type = "AUC/ROC"
    graph_index: int


class ConfusionMatrixChart(Schema):
    name: str

    values: List[List[float]]

    chart_type = "CONFUSION_MATRIX"
    graph_index: int


Chart = Union[TimeSeriesChart, AucRocChart, ConfusionMatrixChart]


class DetailedTrainResultsData(Schema):
    basic_info: TrainResultsData

    all_metrics: List[Chart]